### PR TITLE
Issue439 idp ip discovery

### PR DESCRIPTION
**Pull request by nhoobin**
WIP commit, built on top of PR #444

### CODE REVIEW

File: auth.php [hunk 1]
```diff
@@ -525,8 +533,15 @@ public function saml_login() {
...
-            $idpurl = new moodle_url('/auth/saml2/selectidp.php');
-            redirect($idpurl);
+            // At this stage there is no alias, get-param or default IdP configured.
+            // On a multi-idp system, now check for any whitelisted IP address redirection.
+            $entitiyid = $this->check_whitelisted_ip_redirect();
+            if ($entitiyid !== null) {
+                $SESSION->saml2idp = $entitiyid;
+            } else {
+                $idpurl = new moodle_url('/auth/saml2/selectidp.php');
+                redirect($idpurl);
```
Comment (brendanheywood) [inline auth.php:543]: shouldn't this be a string return not a redirect? + data provider

File: auth.php [hunk 2]
```diff
@@ -706,6 +721,29 @@ protected function handle_blocked_access() {
...
+     * This is used in two places, firstly to determine if a saml redirect is to happen.
+     * Secondly to determine which IdP to force the redirect to.
+     *
+     * @return bool|string
+     */
+    protected function check_whitelisted_ip_redirect() {
+        $found = false;
+        foreach ($this->metadataentities as $idpentities) {
+            foreach ($idpentities as $md5idpentityid => $idpentity) {
+                $list = explode("\n", $idpentity->whitelist);
```
Comment (brendanheywood) [inline auth.php:737]: look at remoteip_in_list() to remove a chunk of this code and also honor things like comments in the ip list

File: classes/form/selectidp_buttons.php [hunk 3]
```diff
@@ -52,6 +52,7 @@ public function definition() {
         $wants = $this->_customdata['wants'];
 
         $mform->addElement('hidden', 'wants', $wants);
+        $mform->setType('wants', PARAM_RAW);
```
Comment (brendanheywood) [inline classes/form/selectidp_buttons.php:55]: should this be PARAM_URL ?

File: classes/form/selectidp_dropdown.php [hunk 4]
```diff
@@ -61,6 +60,7 @@ public function definition() {
         }
 
         $mform->addElement('hidden', 'wants', $wants);
+        $mform->setType('wants', PARAM_RAW);
```
Comment (brendanheywood) [inline classes/form/selectidp_dropdown.php:63]: ditto

File: classes/form/selectidp_buttons.php (hunk 3 above)
Comment (nhoobin) [reply classes/form/selectidp_buttons.php]: Cheers, fixed.

File: classes/form/selectidp_dropdown.php (hunk 4 above)
Comment (nhoobin) [reply classes/form/selectidp_dropdown.php]: Cheers, fixed.

File: auth.php (hunk 2 above)
Comment (nhoobin) [reply auth.php]: I've made some changes for it to use address_in_subnet(), which takes a comma separated list of subnets. Grants greater flexibility than just a list of IPs

File: auth.php (hunk 1 above)
Comment (nhoobin) [reply auth.php:543]: This is the saml_login() function, this has no check for the return type in loginpage_hook(). Want me to return here and add more login in the hook?

File: lang/en/auth_saml2.php [hunk 5]
```diff
@@ -110,6 +110,7 @@
 $string['multiidp:label:defaultidp'] = 'Default IdP';
 $string['multiidp:label:admin'] = 'Admin';
 $string['multiidp:label:whitelist'] = 'Redirected IP addresses';
+$string['multiidp:label:whitelist_help'] = 'If set, it will force clients to this IdP. Format: xxx.xxx.xxx.xxx/bitmask. Separate multiple subnets with \',\' (comma).';
```
Comment (brendanheywood) [inline lang/en/auth_saml2.php:113]: the delimiter is just \n ?

https://github.com/moodle/moodle/blob/master/lib/classes/ip_utils.php#L236

File: tests/phpunit/locallib_test.php [hunk 6]
```diff
@@ -182,6 +182,10 @@ public function should_login_redirect_testcases() {
     public function test_check_whitelisted_ip_redirect($saml, $whitelist, $expected) {
         $this->resetAfterTest();
 
+        // Setting an address here as getremoteaddr() will return default 0.0.0.0 which then is ignored by the address_in_subnet
+        // function.
+        $_SERVER['REMOTE_ADDR'] = '127.0.0.1';
```
Comment (brendanheywood) [inline tests/phpunit/locallib_test.php:187]: I'd move this into the data provider

File: tests/phpunit/locallib_test.php [hunk 7]
```diff
@@ -209,8 +213,8 @@ public function check_whitelisted_ip_redirect_testcases() {
             'saml off, no ip, no redirect'              => ['off', '', false],
             'saml off, getremoteaddr(), no redirect'    => ['off', getremoteaddr(), false],
             'saml not specified, junk, no redirect'     => [null, 'qwer1234!@#$qwer', false],
-            'saml not specified, junk+ip, yes redirect' => [null, 'qwer1234!@#$qwer'."\n".getremoteaddr(), true],
-            'saml not specified, localip, yes redirect' => [null, getremoteaddr()."\n127.0.0.1\n0.0.0.0", true],
+            'saml not specified, junk+ip, yes redirect' => [null, 'qwer1234!@#$qwer,127.0.0.1', true],
+            'saml not specified, localip, yes redirect' => [null, '127.0.0.,1.2.3.4', true],
```
Comment (brendanheywood) [inline tests/phpunit/locallib_test.php:217]: see above, pretty sure this should be a \n not a comma

File: lang/en/auth_saml2.php (hunk 5 above)
Comment (nhoobin) [reply lang/en/auth_saml2.php:113]: Ah. I Was using address_in_subnet(), not \core\ip_utils::is_ip_in_subnet_list, I'll adjust that.

File: tests/phpunit/locallib_test.php (hunk 7 above)
Comment (nhoobin) [reply tests/phpunit/locallib_test.php]: By moving to using ip_utils::is_ip_in_subnet_list, updated tests to reflect new lines

File: tests/phpunit/locallib_test.php (hunk 6 above)
Comment (nhoobin) [reply tests/phpunit/locallib_test.php]: Moved into data provider. Additionally there is an $active flag too, review change in check_whitelisted_ip_redirect()

File: auth.php (hunk 1 above)
Comment (brendanheywood) [reply auth.php:543]: ah yup I got mixed up, leave this as is

**Pull request merged**